File: 36-valid_sudoku.rb
def is_valid_sudoku(board)
    # Check rows
    board.each do |row|
        return false if has_duplicates(row)
    end
    
    # Check columns
    9.times do |i|
        column = []
        9.times do |j|
            column << board[j][i]
        end
        return false if has_duplicates(column)
    end
    
    # Check sub-boxes
    3.times do |i|
        3.times do |j|
            sub_box = []
            3.times do |k|
                3.times do |l|
                    sub_box << board[3*i+k][3*j+l]
                end
            end
            return false if has_duplicates(sub_box)
        end
    end
    
    # If all checks pass, the board is valid
    return true
end

def has_duplicates(arr)
    # Use a hash table to keep track of digits seen so far
    seen = {}
    arr.each do |digit|
        if digit != '.'
            return true if seen[digit]
            seen[digit] = true
        end
    end
    return false
end

File: 36-valid_sudoku.py
from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        def is_valid_block(block):
            block = [num for num in block if num != '.']
            return len(block) == len(set(block))
        
        # Check rows
        for row in board:
            if not is_valid_block(row):
                return False
        
        # Check columns
        for col in zip(*board):
            if not is_valid_block(col):
                return False
        
        # Check 3x3 sub-boxes
        for i in range(0, 9, 3):
            for j in range(0, 9, 3):
                block = [board[x][y] for x in range(i, i+3) for y in range(j, j+3)]
                if not is_valid_block(block):
                    return False
        
        return True
